lib/ansible/module_utils/common/arg_spec.py

```diff
@@ -0,0 +1,125 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2020 Ansible Project
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+from copy import deepcopy
+
+from ansible.module_utils.common._collections_compat import (
+    Sequence,
+)
+
+from ansible.module_utils.common.parameters import (
+    get_unsupported_parameters,
+    handle_aliases,
+    list_no_log_values,
+    remove_values,
+    set_defaults,
+    set_fallbacks,
+    validate_argument_types,  # Rename this because it actually does coercion?
+    validate_argument_values,
+    validate_sub_spec,
+)
+
+from ansible.module_utils.common.text.converters import to_native
+from ansible.module_utils.common.warnings import warn
+from ansible.module_utils.common.validation import (
+    check_required_arguments,
+)
+
+from ansible.module_utils.six import string_types
+
+
+class ArgumentSpecValidator():
+    """Argument spec validation class"""
+
+    def __init__(self, argument_spec, parameters):
+        self._error_messages = []
+        self._no_log_values = set()
+        self.argument_spec = argument_spec
+        # Make a copy of the original parameters to avoid changing them
+        self._validated_parameters = deepcopy(parameters)
+        self._unsupported_parameters = set()
+
+    @property
+    def error_messages(self):
+        return self._error_messages
+
+    @property
+    def validated_parameters(self):
+        return self._validated_parameters
+
+    def _add_error(self, error):
+        if isinstance(error, string_types):
+            self._error_messages.append(error)
+        elif isinstance(error, Sequence):
+            self._error_messages.extend(error)
+        else:
+            raise ValueError('Error messages must be a string or sequence not a %s' % type(error))
+
+    def _sanitize_error_messages(self):
+        self._error_messages = remove_values(self._error_messages, self._no_log_values)
+
+    def validate(self):
+        """Validate module parameters against argument spec.
+
+        :Example:
+
+        [good example goes here]
+
+        :param argument_spec: Specification of parameters, type, and valid values
+        :type argument_spec: dict
+
+        :param parameters: Parameters provided to the role
+        :type parameters: dict
+
+        :returns: True if no errors were encountered, False if any errors were encountered.
+        :rtype: bool
+        """
+
+        self._no_log_values.update(set_fallbacks(self.argument_spec, self._validated_parameters))
```

Comment:
set_fallbacks is already called in main code as well as update to no_log, do we really need it inside validate again?
values should be set before calling validation, not inside of it.